File: java/dynamodb/src/main/java/UserStore.java
import com.amazonaws.services.dynamodbv2.AmazonDynamoDB;
import com.amazonaws.services.dynamodbv2.model.*;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public class UserStore {
    static final String USER_TABLE = "users";

    public static class UserEntry {
        String userId;
        String apiKey;

        public UserEntry(String userId, String apiKey) {
            this.userId = userId;
            this.apiKey = apiKey;
        }
    }

    private AmazonDynamoDB client;

    public UserStore(AmazonDynamoDB client) {
        this.client = client;
    }

    UserEntry getUser(String userId) {
        HashMap<String, AttributeValue> keyToGet = new HashMap<>();
        keyToGet.put("userId", new AttributeValue(userId));

        GetItemResult result = client.getItem(new GetItemRequest(USER_TABLE, keyToGet));
        Map<String, AttributeValue> item = result.getItem();

        return new UserEntry(item.get("userId").getS(), item.get("apiKey").getS());
    }

    List<UserEntry> listUsers() {
        ScanResult result = client.scan(new ScanRequest(USER_TABLE));
        List<Map<String, AttributeValue>> items = result.getItems();

        List<UserEntry> output = new ArrayList<>();
        for (Map<String, AttributeValue> item : items) {
            output.add(new UserEntry(item.get("userId").getS(), item.get("apiKey").getS()));
        }
        return output;
    }

    List<UserEntry> getUserWithApiKey(String apiKey) {
        HashMap<String, AttributeValue> attrValues = new HashMap<>();
        attrValues.put(":apiKey", new AttributeValue(apiKey));

        QueryRequest req = new QueryRequest(USER_TABLE)
                .withIndexName("apiKey-index")
                .withKeyConditionExpression("apiKey = :apiKey")
                .withExpressionAttributeValues(attrValues);

        QueryResult result = client.query(req);
        List<Map<String, AttributeValue>> items = result.getItems();

        List<UserEntry> output = new ArrayList<>();
        for (Map<String, AttributeValue> item : items) {
            output.add(new UserEntry(item.get("userId").getS(), item.get("apiKey").getS()));
        }
        return output;
    }

    void deleteUser(String userId) {
        HashMap<String, AttributeValue> keyToGet = new HashMap<>();
        keyToGet.put("userId", new AttributeValue(userId));
        client.deleteItem(new DeleteItemRequest(USER_TABLE, keyToGet));
    }

    void storeUser(UserEntry user) {
        HashMap<String, AttributeValue> item = new HashMap<>();
        item.put("userId", new AttributeValue(user.userId));
        item.put("apiKey", new AttributeValue(user.apiKey));
        client.putItem(new PutItemRequest(USER_TABLE, item));
    }
}

File: python/dynamodb/user_store.py


class UserStore:
    def __init__(self, dynamodb_client):
        self.dynamodb_client = dynamodb_client

    def get_user(self, user_id):
        table = self.dynamodb_client.Table('users')
        response = table.get_item(
            Key={
                'userId': user_id
            }
        )
        if 'Item' in response:
            return response['Item']
        return None

    def delete_user(self, user_id):
        table = self.dynamodb_client.Table('users')
        response = table.delete_item(
            Key={
                'userId': user_id
            }
        )
        return response

    def put_user(self, user_id, api_key, configs):
        table = self.dynamodb_client.Table('users')
        response = table.put_item(
            Item={
                'userId': user_id,
                'apiKey': api_key,
                'configs': configs
            }
        )
        return response

    def list_users(self):
        table = self.dynamodb_client.Table('users')
        response = table.scan()
        return response['Items']